ddtrace/internal/utils/time.py

```diff
@@ -101,3 +101,37 @@ def stop(self):
             raise RuntimeError("Can not stop a stopwatch that has not been" " started")
         self._stopped_at = compat.monotonic()
         return self
+
+
+class HourGlass(object):
+    """An implementation of an hourglass."""
+
+    def __init__(self, duration):
+        # type: (float) -> None
+        t = compat.monotonic()
+        self._duration = duration
+        self._started_at = t - duration
+        self._end_at = t
+
+    def turn(self):
+        # type: () -> None
+        """Turn the hourglass."""
+        t = compat.monotonic()
+        top_0 = self._end_at - self._started_at
+        bottom = self._duration - top_0 + min(t - self._started_at, top_0)
+
+        self._started_at = t
+        self._end_at = t + bottom
+
+    def trickling(self):
+        # type: () -> bool
+        """Check if sand is still trickling."""
+        return compat.monotonic() < self._end_at
```

Comment:
what is cost of calling `compat.monotonic()` ? we might want to cache trickling as once this is true it should never change.

Reply:
monotonic is implemented in native code, so should be quite cheap